### ISSUE TITLE
Generating schema with an Array property allowing multiple types of an interface

### ISSUE THREAD

**Issue by HolyOne** [gregsdennis/json-everything#336]
Hello, I am strugling to programaticly generate schema for my configuration and I wondered if there is a practicle way I am missing.

My Example Config file is this

```json
{
   "$schema": "config-schema.json" ,
    "BackupDir": "c:\\backup\\",
    "CompressionLevel": 2,
    "Items": [
        {"Type": "File" , "Path" : "c:\\SomeDirectory", "ArchiveSize":128 },
        {"Type": "File" , "Path" : "c:\\SomeDirectory2", "ArchiveSize":128 },
        {"Type": "MySQL" , "Server" : "localhost", "DB":"mydata"  },
        {"Type": "MySQL" , "Server" : "localhost", "DB":"mydb" },
        {"Type": "MSSQL" , "Server" : "localhost", "DB":"mymsdb", "Username":"sa", "Password":"...." }
    ]
 }
```

As you see In c#, I have an Items property of IEnumerable<IBackupInfo> which only requires "Type" Parameter. 
But It may also have other properties because Items can contain any class inherited from IBackupInfo.

```c#
    public interface IBackupInfo
	{
        string Type { get;   }
    }  
```

I read the Items, cast them to the appropriate class using the common "Type" property. 

How can I programaticly generate the schema file for this, (which is aware of all the class types i provide) ?

**Comment by gregsdennis** [comment]
The generation system would need to scan the loaded assemblies for implementations of an interface and then generate a `oneOf` or `anyOf`.  Currently it doesn't do this.

I suggest generating each of your implementation types separately and referencing them in an `anyOf` manually, using what you get from generating your config schema as a starting point.

**Comment by gregsdennis** [comment]
I'm looking into implementing something for this.

It does look like interface types generate appropriate schemas.  The missing piece is that there isn't any generation for _implementations_ of the interface, so no extra properties are added.

To expand on your example,

```c#
public interface IBackupInfo
{
    string Type { get; }
}

public class FileBackupInfo : IBackupInfo
{
    public string Type { get; }
    public string Path { get; }
    public int ArchiveSize { get; }
}

public class MySqlBackupInfo : IBackupInfo
{
    public string Type { get; }
    public string Server { get; }
    public string DB { get; }
    public string? Username { get; }
    public string? Password { get; }
}
```

an `IEnumerable<IBackupInfo>` should generate

```json
{
  "type": "array",
  "items": {
    {
      "type": "object",
      "properties": {
        "Type": { "type": "string" }
      },
      "required": [ "Type" ]
    }
  }
}
```

Note that only the interface definition is included.  From what I understand of your post, you also want `FileBackupInfo` and `MySqlBackupInfo` to be included in the schema, correct?

I would like to propose the following generation:

```json
{
  "$defs": {
    "IBackupInfo": {
      {
        "type": "object",
        "properties": {
          "Type": { "type": "string" }
        },
        "required": [ "Type" ]
      }
    },
    "FileBackupInfo": {
      {
        "$ref": "#/$defs/IBackupInfo",
        "type": "object",
        "properties": {
          "Path": { "type": "string" },
          "ArchiveSize": { "type": "integer" }
        },
        "required": [ "Path", "ArchiveSize" ],
        "unevaluatedProperties": false
      }
    },
    "MySqlBackupInfo": {
      {
        "$ref": "#/$defs/IBackupInfo",
        "type": "object",
        "properties": {
          "Server": { "type": "string" },
          "DB": { "type": "string" },
          "Username": { "type": [ "string", "null" ] },
          "Password": { "type": [ "string", "null" ] },
        },
        "required": [ "Server", "DB" ],
        "unevaluatedProperties": false
      }
    }
  },
  "type": "array",
  "items": {
    "oneOf": [
      { "$ref": "#/$defs/FileBackupInfo" },
      { "$ref": "#/$defs/MySqlBackupInfo" }
    ]
  }
}
```

Basically, it creates definitions for each of the types and references thsoe definitions in a `oneOf`.

There may be some additional requirements like adding some more attributes to the class definitions, but this would be the basic gist of it.

Does this sound like what you're going for?

**Comment by HolyOne** [comment]
Thank you very much for your answer. I did it by manually adding anyof's
but I will work on your last post it seems like a better solution than mine

**Comment by gregsdennis** [comment]
> I did it by manually adding anyof's

This is actually the preferred goal for this library.  It's not intended as a production-ready tool.  I don't believe _any_ code generation tool like this can be production-ready.  This is a dev-time tool; not a run-time tool.  It's vitally important that developers review code that's generated to ensure that it expresses their intent.  That's why I have this disclaimer at the top of the documentation:

> ### IMPORTANT
> Ideally, this functionality should be used to create a starting point in authoring a schema. The schemas output by this library should be reviewed by actual people prior to being put into a production system.
> 
> **The JSON Schema team generally recommends against real-time schema generation.**
> 
> The suggested approach is to have a dedicated schema generation data model, separate from your domain model and DTOs, from which schemas are generated. Generation can be done as an independent developer activity or as part of a CI/CD build chain. This helps separate serialization and validation concerns and prevent overburdening your domain/DTO models with a bunch of attributes.
